src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -0,0 +1,470 @@
+                    break;
+
+                case TarEntryType.RegularFile:
+                case TarEntryType.V7RegularFile:
+                case TarEntryType.ContiguousFile:
+                    ExtractAsRegularFile(filePath);
+                    break;
+
+                case TarEntryType.SymbolicLink:
+                    Debug.Assert(!string.IsNullOrEmpty(linkTargetPath));
+                    FileInfo link = new(filePath);
+                    link.CreateAsSymbolicLink(linkTargetPath);
+                    break;
+
+                case TarEntryType.HardLink:
+                    Debug.Assert(!string.IsNullOrEmpty(linkTargetPath));
+                    ExtractAsHardLink(linkTargetPath, filePath);
+                    break;
+
+                case TarEntryType.BlockDevice:
+                    ExtractAsBlockDevice(filePath);
+                    break;
+
+                case TarEntryType.CharacterDevice:
+                    ExtractAsCharacterDevice(filePath);
+                    break;
+
+                case TarEntryType.Fifo:
+                    ExtractAsFifo(filePath);
+                    break;
+
+                case TarEntryType.ExtendedAttributes:
+                case TarEntryType.GlobalExtendedAttributes:
+                case TarEntryType.LongPath:
+                case TarEntryType.LongLink:
+                    Debug.Assert(false, $"Metadata entry type should not be visible: '{EntryType}'");
+                    break;
+
+                case TarEntryType.MultiVolume:
+                case TarEntryType.RenamedOrSymlinked:
+                case TarEntryType.SparseFile:
+                case TarEntryType.TapeVolume:
+                default:
+                    throw new InvalidOperationException(string.Format(SR.TarEntryTypeNotSupportedForExtracting, EntryType));
+            }
+        }
+
+        // Verifies if the specified paths make sense for the current type of entry.
+        private void VerifyPathsForEntryType(string filePath, string? linkTargetPath, bool overwrite)
+        {
+            string? directoryPath = Path.GetDirectoryName(filePath);
+            // If the destination contains a directory segment, need to check that it exists
+            if (!string.IsNullOrEmpty(directoryPath) && !Path.Exists(directoryPath))
+            {
+                throw new IOException(string.Format(SR.IO_PathNotFound_NoPathName, filePath));
+            }
+
+            if (!Path.Exists(filePath))
+            {
+                return;
+            }
+
+            // We never want to overwrite a directory, so we always throw
+            if (Directory.Exists(filePath))
+            {
+                throw new IOException(string.Format(SR.IO_AlreadyExists_Name, filePath));
+            }
+
+            // A file exists at this point
+            if (!overwrite)
+            {
+                throw new IOException(string.Format(SR.IO_AlreadyExists_Name, filePath));
+            }
+            File.Delete(filePath);
```

Comment:
why would we like to remove the file now, if the method can still throw few checks later?

If we want to overwrite a file we can use `FileMode.Create` when creating a `FileStream`:

> Specifies that the operating system should create a new file. If the file already exists, it will be overwritten.